FHEM/14_SD_UT.pm

```diff
@@ -1360,6 +1385,12 @@ sub SD_UT_Parse {
         $name = $model.'_' . $deviceCode;
       }
     }
+    ### BeSmart_S4 [P78] ###
+    if (!$def && $protocol == 78) {
+      $deviceCode = substr($rawData,0,3);
+      $devicedef = 'BeSmart_S4 ' . $deviceCode;
+      $def = $modules{SD_UT}{defptr}{$devicedef};
+    }
     ### NEFF SF01_01319004 || BOSCH SF01_01319004_Typ2 [P86] ###
     if (!$def && $protocol == 86) {
```

Comment:
Hier wäre dann ein elsif passend denke ich.

Reply:
Sieht so aus, habe ich geändert.

Reply:
Ich musste die Änderung wieder zurück nehmen. Damit funktionierte die Fernbedienung TR_502MSV Protokoll 34 nicht mehr.

Reply:
Puhh, könnte sein, dass ich das schon mal angemerkt habe. Kann aber auch in einem anderen Modul gewesen sein.

Hatte gestern nur die paar Zeilen gesehen und an gesehen, dass in den folgenden Abfragen mit esleif gearbeitet wird.
In den darüber liegenden Zeilen wird mit IF (!def usw. gearbeitet.

Das ist schon höchst seltsam. Zum einen wird geprüft ob $def unwahr ist. Eigentlich will man aber wissen ob es defined ist oder nicht. Ich verstehe aber copy & paste. 
Die Vielen IF Abfragen kommen nach meinem Verständnis daher, da nicht sicher ist, ob in einem der IF Blöcke auch eine Definition gefunden wird. Daher muss man das so lange durchprobieren, bis man was passendes gefunden hat.

Was dabei allerdings verwirrt, dass ja auch $protocol berücksichtigt wird und das kann nur einen der Werte haben.
Um hier mal eine Idee anzubringen, wäre es vermutlich klarer was passiert, wenn die IF-Abfragen
1284-1395 in eigene codeblöcke zu verlagern die mit einem Return beendet werden können.

Eine einfache Variante könnten Subs sein, vielleicht klappt es auch mit einer Schleife.

Anstelle mehrere
```
    if (!$def && $protocol == 78) {
      $deviceCode = substr($rawData,0,3);
      $devicedef = 'BeSmart_S4 ' . $deviceCode;
      $def = $modules{SD_UT}{defptr}{$devicedef};
    }
``` 
hintereinander zu hängen könnte man dann sowas schreiben:
```
    if ($protocol == 78) {
      $deviceCode = substr($rawData,0,3);
      $devicedef = 'BeSmart_S4 ' . $deviceCode;
      return $modules{SD_UT}{defptr}{$devicedef}; if defined $modules{SD_UT}{defptr}{$devicedef};
    }
```

Wäre halt mal so ne Idee, nicht alle IF Prüfungen durcharbeiten zu müssen.
Alternativ ginge es auch mit einem codeblock und do statements 

```
getDef: {
		$protocol ==78 && do { $deviceCode = substr($rawData,0,3);  $devicedef = 'BeSmart_S4 ' . $deviceCode; $ptr = if defined $modules{SD_UT}{defptr}{$devicedef};);
		$protocol ==68 && do { ... }
	}
```

Da der Code in dem do blöcken fast immer gleich ist, könnte das natürlich ebenfalls in eine kleine sub ausgelagert werden:

```
sub _try_get_defptr
{
  my $rawData = shift // return;
  my $start = shift // 0;
  my $end = shift // length $rawData;
  my $nameIdentifier = shift // return;

  return if (length $rawData < $end);
  return if (length $rawData < $start);

  $deviceCode = substr($rawData,0,3);
  return $modules{SD_UT}{defptr}{$nameIdentifier.$deviceCode} if  defined($modules{SD_UT}{defptr}{$nameIdentifier.$deviceCode});
 return ;
}
``` 
Der Teil mit dem Do würde sich dann in etwa so darstellen.
``` 
   $protocol ==78 && do {  my $ptr = _try_get_defptr($rawdata,0,3,'BeSmart_S4'); last if defined($ptr) };
``` 

Gut, das waren jetzt ganz viele Ideen, wie man die IF - Schlage hier ein bissl in den Griff bekommen könnte.
Hat alles aber nichts mit diesem PR zu tun sondern eher, wie der code und das Ergebnis lesbarer werden könnten.

Denn ich glaube der Fehler, wieso das mit dem IF und Else nicht ghelappt hat, liegt  Zeile 1736.
Das elseif ist hier vermutlich falsch, denn in den vielen IF (!def) darüber wird ja versucht für $def einen Wert zu suchen.
Ab 1736 geht es darum den $devicecode zu definieren. Quasi eine neue IF- Elseif schlange :)

Reply:
Danke für die ausführliche Erläuterung.
Wir werden uns mal mit den Vorschlägen vertraut machen um da vielleicht den Code übersichtlicher zu gestalten.

Reply:
Vielleicht könnt ihr meine Annahme zu dem Fehler bestätigen oder wiederlegen?

Reply:
>Vielleicht könnt ihr meine Annahme zu dem Fehler bestätigen oder wiederlegen?

Machen wir doch gern. Den Fehler können wir nicht bestätigen. Zusammenfassend haben wir deine Punkte mal aufgegriffen nachdem wir alles uns angesehen haben.

>...dass ja auch $protocol berücksichtigt wird und das kann nur einen der Werte haben.

Das ist falsch.

1)
```
    ### Westinghouse Buttons_five [P29] ###
    if (!$def && ($protocol == 29 || $protocol == 30)) {

    ### Novy_840029 [P86] ###
    if (!$def && ($protocol == 86 || $protocol == 81)) {
```
Das kommt zu tragen, da manche Protokolle sehr ähnlich sein können und aufgrund der Toleranzen "doppelt anschlagen" bei einem Empfang.

2.
Verschiedene Remotes das gleiche Prot verwenden.
```
  if($hlen == 3 || $hlen == 5) {
    ### Novy_840029 [P86] ###
    if (!$def && ($protocol == 86 || $protocol == 81)) {
      $deviceCode = substr($rawData,0,2);
      $devicedef = 'Novy_840029 ' . $deviceCode;
      $def = $modules{SD_UT}{defptr}{$devicedef};
    }
    ### Novy_840039 [P86] ###
    if (!$def && ($protocol == 86 || $protocol == 81)) {
      $deviceCode = substr($rawData,0,2);
      $devicedef = 'Novy_840039 ' . $deviceCode;
      $def = $modules{SD_UT}{defptr}{$devicedef};
    }
  }
```

>... 1284-1395 in eigene codeblöcke zu verlagern die mit einem Return beendet werden können.

Das bringt keine Vorteile weil:
- $deviceCode = substr($rawData,0,3) unterschiedlich ist
- $protocol NICHT immer ein Werte annehmen kann
- $devicedef unterschiedlich ist je Remote / Typ
- $deviceCode unterschiedlich ist je Remote / Typ

> ... ebenfalls in eine kleine sub ausgelagert 

Das bringt keine Vorteile, da dann 2 Varianten der Schreibweise im Modul verwendet werden müssen, da keine Vereinheitlichtung möglich ist.
Sollte man es dennoch umschreiben, so MEHRT sich der Code mit vielen Subs und es quellt auf. (undurchsichtiger) :-(

> ... Denn ich glaube der Fehler, wieso das mit dem IF und Else nicht geklappt hat, liegt Zeile 1736.

Nein es lag an der falschen Schreibweise ELSIF.

Nun hoffe ich, dir ein wenig mehr Licht ins Dunkle gebracht zu haben. Deine Überlegungen sind richtig, aber diese sind hier nicht praktikabel an der Stelle.